python/Ganga/test/GPI/FileTests/TestMassStorageGetPut.py

```diff
@@ -0,0 +1,141 @@
+    MassStorageTestConfig = {'defaultProtocol': 'file://',
+                             'fileExtensions': [''],
+                             'uploadOptions': {'path': outputFilePath, 'cp_cmd': 'cp', 'ls_cmd': 'ls', 'mkdir_cmd': 'mkdir -p'},
+                             'backendPostprocess': {'LSF': 'client', 'LCG': 'client', 'ARC': 'client', 'Dirac': 'client',
+                                                    'PBS': 'client', 'Interactive': 'client', 'Local': 'client', 'CREAM': 'client'}}
+
+    def setUp(self):
+        """
+        Configure the MassStorageFile for the test
+        """
+        extra_opts=[('PollThread', 'autostart', 'False'),
+                    ('Local', 'remove_workdir', 'False'),
+                    ('TestingFramework', 'AutoCleanup', 'False'),
+                    ('Output', 'MassStorageFile', self.MassStorageTestConfig),
+                    ('Output', 'FailJobIfNoOutputMatched', 'True')]
+        super(TestMassStorageGetPut, self).setUp(extra_opts=extra_opts)
+
+    @classmethod
+    def cleanUp(cls):
+        """ Cleanup the current temp objects """
+
+        from Ganga.GPI import jobs
+        for j in jobs:
+            shutil.rmtree(j.backend.workdir, ignore_errors=True)
+            j.remove()
+
+        for file_ in cls._temp_files:
+            os.unlink(file_)
+        cls._temp_files = []
+
+        for file_ in cls._managed_files:
+            os.unlink(os.path.join(cls.outputFilePath, file_.namePattern))
+        cls._managed_files = []
+
+        shutil.rmtree(cls.outputFilePath, ignore_errors=True)
+
+    def test_a_test_put(self):
+        """Test that a job can be submitted with inputfiles in the input"""
+
+        from Ganga.GPI import MassStorageFile
+
+        _ext = '.root'
+        file_1 = generate_unique_temp_file(_ext)
+        file_2 = generate_unique_temp_file(_ext)
+        self._temp_files.append(file_1)
+        self._temp_files.append(file_2)
+        msf_1 = MassStorageFile(file_1)
+        msf_2 = MassStorageFile(file_2)
+        self._managed_files.append(msf_1)
+        self._managed_files.append(msf_2)
+        msf_1.put()
+        msf_2.put()
+
+        for file_ in [msf for msf in (msf_1, msf_2)]:
+            assert os.path.isfile(os.path.join(self.outputFilePath, file_.namePattern))
+            file_.localDir = ''
+            assert file_.localDir == ''
+
+    def test_b_test_get(self):
+        """Test that the files were made accessible to the WN area and collected as LocalFile objects in outputfiles"""
+
+        from Ganga.GPIDev.Base.Proxy import stripProxy
+        from Ganga.GPI import Job
+
+        tmpdir = '/tmp/testMassStorageGet'
+        if not os.path.isdir(tmpdir):
+            os.makedirs(tmpdir)
+
+        # Test in the case that the files don't have a parent or a localDir
+        for file_ in self._managed_files:
+            file_.localDir = ''
+            failed = True
```

Comment:
Sure. Unlike C/C++/many others, Python does not have local scoping of variables limited to within blocks. So:
```python
if foo:
    bar = 6
assert bar == 6
```
will work (as long as `foo == True`) even though you might think that `bar` was only declared in the scope of the `if` statement.

You can take this a step further and take the following example:
```python
if foo:
    bar = 6
else:
    bar = 10
assert bar > 4
```
This will always work as `bar` is in scope and you know that whichever path the code took, it will always have given a value to `bar`.

Therefore, you can do whatever you want to `bar` before the `if` blocks and it will do nothing:
```python
bar = 0
if foo:
    bar = 6
else:
    bar = 10
assert bar > 4
```
as `bar` will *always* be given a value inside the `if`/`else` blocks.

Likewise,
```python
bar = 'not a bool'
try:
    ...any code here which may or may not raise an exception of any kind
    bar = False
except GangaException:
    bar = True

assert isinstance(bar, bool)
```
will *always* work as you know that either:

1. The `try` block got to the end and therefore `bar == False`
2. A `GangaException` was raised before it could finish and so `bar == True`
3. Any other exception was raise in which case the stack will be unwound and the code will never get to the `assert`.

This means that there's no point in setting `bar` to anything before the `try`/`except` as:

1. It's not needed for scoping
2. Its value will always be overwritten